### ISSUE TITLE
AppService: Changing AppSettings after initial launch throws error

### ISSUE THREAD

**Issue by MitchellGerdisch** [pulumi/pulumi-azure#740]
### Set up:
Use one of these azure app service examples:
(Although probably any of the azure-XX-appservice examples will show the same issue.)
- [azure-cs-appservice](https://github.com/pulumi/examples/tree/master/azure-cs-appservice), or
- [azure-ts-appservice](https://github.com/pulumi/examples/tree/master/azure-ts-appservice)

Launch the stack.
- Go to Azure console, select the resource group that was created by the stack, select the "App Service" that was created by the stack.
- In the console panel along the left, find Settings/Configuration and select that.
- In the Configuration view, you will see a section named "Application settings" and in that section you will see the 4 settings created by the pulumi code. (e.g. APPINSIGHTS_INSTRUMENTATIONKEY, APPLICATIONINSIGHTS_CONNECTION_STRING, etc).

### How to cause the error
- Go to the pulumi code and add an AppSetting, it can be any key:value pair, so for example add 
`MY_NEW_SETTING: "new setting`
- `pulumi up` to apply the change

This will cause an error like that shown at the bottom.

However, if you refresh the azure console for the App Service (click the refresh button and the continue button in the console), you will see the new setting was added.
Furthermore, if you had made other changes as part of the update, and those changes were not executed before this error, those changes are not made.
Any `pulumi up` from this point forward will fail until you do a `pulumi refresh`.

### WORK AROUND
Run `pulumi refresh` after the error.
This will synchronize the AppSettings and allow subsequent updates (for non-AppSettings) to work.
Subsequent AppSettings changes will, naturally, throw the error on `pulumi up` and will require another `pulumi refresh`

###  pulumi up error after changing AppSettings 
```
Diagnostics:
  azure:appservice:AppService (mitchts-as):
    error: transport is closing
 
  pulumi:pulumi:Stack (azure-appservice-dev):
    panic: interface conversion: interface {} is nil, not map[string]interface {}
    goroutine 48 [running]:
    github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/web.expandAppServiceLogs(0x5310fe0, 0xc0014cb3e0, 0x4, 0x5310fe0, 0xc0014cb3e0, 0x1)
        /home/runner/go/pkg/mod/github.com/pulumi/terraform-provider-azurerm@v1.41.1-0.20201016142117-857b1cde95f3/azurerm/internal/services/web/app_service.go:1382 +0xcad
    github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/web.resourceArmAppServiceUpdate(0xc000a76770, 0x549e2c0, 0xc000164b00, 0x0, 0x0)
        /home/runner/go/pkg/mod/github.com/pulumi/terraform-provider-azurerm@v1.41.1-0.20201016142117-857b1cde95f3/azurerm/internal/services/web/resource_arm_app_service.go:487 +0x1c40
    github.com/hashicorp/terraform-plugin-sdk/helper/schema.(*Resource).Apply(0xc0006e9950, 0xc0005e8550, 0xc0015be300, 0x549e2c0, 0xc000164b00, 0x1, 0xffffffffffffffff, 0x0)
        /home/runner/go/pkg/mod/github.com/hashicorp/terraform-plugin-sdk@v1.13.1/helper/schema/resource.go:316 +0x263
    github.com/hashicorp/terraform-plugin-sdk/helper/schema.(*Provider).Apply(0xc00019cd80, 0xc0019375e0, 0xc0005e8550, 0xc0015be300, 0x0, 0x0, 0x0)
        /home/runner/go/pkg/mod/github.com/hashicorp/terraform-plugin-sdk@v1.13.1/helper/schema/provider.go:294 +0x99
    github.com/pulumi/pulumi-terraform-bridge/v2/pkg/tfshim/sdk-v1.v1Provider.Apply(0xc00019cd80, 0x5c05520, 0x13, 0x65cddc0, 0xc0005e8550, 0x65e8be0, 0xc0015be300, 0x0, 0x65e8be0, 0xc0015be300, ...)
        /home/runner/go/pkg/mod/github.com/pulumi/pulumi-terraform-bridge/v2@v2.11.0/pkg/tfshim/sdk-v1/provider.go:97 +0xe5
    github.com/pulumi/pulumi-terraform-bridge/v2/pkg/tfbridge.(*Provider).Update(0xc000a4e380, 0x65ca540, 0xc0008c0930, 0xc0011a2000, 0xc000a4e380, 0x550e201, 0xc000bea300)
        /home/runner/go/pkg/mod/github.com/pulumi/pulumi-terraform-bridge/v2@v2.11.0/pkg/tfbridge/provider.go:957 +0x959
    github.com/pulumi/pulumi/sdk/v2/proto/go._ResourceProvider_Update_Handler.func1(0x65ca540, 0xc0008c0930, 0x5911d40, 0xc0011a2000, 0x58f92e0, 0xa074d00, 0x65ca540, 0xc0008c0930)
        /home/runner/go/pkg/mod/github.com/pulumi/pulumi/sdk/v2@v2.11.3-0.20201009201355-249140242ebb/proto/go/provider.pb.go:2280 +0x89
    github.com/grpc-ecosystem/grpc-opentracing/go/otgrpc.OpenTracingServerInterceptor.func1(0x65ca540, 0xc0008c0000, 0x5911d40, 0xc0011a2000, 0xc001926240, 0xc001926260, 0x0, 0x0, 0x654c7c0, 0xc00027f840)
        /home/runner/go/pkg/mod/github.com/grpc-ecosystem/grpc-opentracing@v0.0.0-20180507213350-8e809c8a8645/go/otgrpc/server.go:57 +0x2eb
    github.com/pulumi/pulumi/sdk/v2/proto/go._ResourceProvider_Update_Handler(0x59c8e80, 0xc000a4e380, 0x65ca540, 0xc0008c0000, 0xc00141e000, 0xc000549760, 0x65ca540, 0xc0008c0000, 0xc0000bc600, 0x12b5)
        /home/runner/go/pkg/mod/github.com/pulumi/pulumi/sdk/v2@v2.11.3-0.20201009201355-249140242ebb/proto/go/provider.pb.go:2282 +0x14b
    google.golang.org/grpc.(*Server).processUnaryRPC(0xc000a4e1c0, 0x6611320, 0xc000782900, 0xc0001aa100, 0xc001132c90, 0xa0378b8, 0x0, 0x0, 0x0)
        /home/runner/go/pkg/mod/google.golang.org/grpc@v1.30.0/server.go:1171 +0x50a
    google.golang.org/grpc.(*Server).handleStream(0xc000a4e1c0, 0x6611320, 0xc000782900, 0xc0001aa100, 0x0)
        /home/runner/go/pkg/mod/google.golang.org/grpc@v1.30.0/server.go:1494 +0xccd
    google.golang.org/grpc.(*Server).serveStreams.func1.2(0xc000dcc050, 0xc000a4e1c0, 0x6611320, 0xc000782900, 0xc0001aa100)
        /home/runner/go/pkg/mod/google.golang.org/grpc@v1.30.0/server.go:834 +0xa1
    created by google.golang.org/grpc.(*Server).serveStreams.func1
        /home/runner/go/pkg/mod/google.golang.org/grpc@v1.30.0/server.go:832 +0x204
 
    error: update failed
```

**Comment by khellang** [comment]
This error is killing me 😭 Even a refresh doesn't help. I'm pretty much stuck.

```
Diagnostics:
  pulumi:pulumi:Stack (srm-dev):
    panic: interface conversion: interface {} is nil, not map[string]interface {}
    goroutine 449 [running]:
    github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/web.expandAppServiceLogs(0x47c1120, 0xc001e41260, 0x4, 0x47c1120, 0xc001e41260, 0x1)
        /home/runner/go/pkg/mod/github.com/pulumi/terraform-provider-azurerm@v1.41.1-0.20201016142117-857b1cde95f3/azurerm/internal/services/web/app_service.go:1382 +0xcb4
    github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/web.resourceArmAppServiceUpdate(0xc000d72770, 0x494d740, 0xc0002c8580, 0x0, 0x0)
        /home/runner/go/pkg/mod/github.com/pulumi/terraform-provider-azurerm@v1.41.1-0.20201016142117-857b1cde95f3/azurerm/internal/services/web/resource_arm_app_service.go:487 +0x1c47
    github.com/hashicorp/terraform-plugin-sdk/helper/schema.(*Resource).Apply(0xc001573560, 0xc0007bf180, 0xc00190b940, 0x494d740, 0xc0002c8580, 0x1, 0x0, 0x0)
        /home/runner/go/pkg/mod/github.com/hashicorp/terraform-plugin-sdk@v1.13.1/helper/schema/resource.go:316 +0x26a
    github.com/hashicorp/terraform-plugin-sdk/helper/schema.(*Provider).Apply(0xc000718c00, 0xc0022175e8, 0xc0007bf180, 0xc00190b940, 0x0, 0x0, 0x0)
        /home/runner/go/pkg/mod/github.com/hashicorp/terraform-plugin-sdk@v1.13.1/helper/schema/provider.go:294 +0xa0
    github.com/pulumi/pulumi-terraform-bridge/v2/pkg/tfshim/sdk-v1.v1Provider.Apply(0xc000718c00, 0x50b3c6a, 0x13, 0x5a6b320, 0xc001f26830, 0x5aada80, 0xc00190b940, 0x0, 0x5aada80, 0xc00190b940, ...)
        /home/runner/go/pkg/mod/github.com/pulumi/pulumi-terraform-bridge/v2@v2.11.1-0.20201020163502-64cff1e50894/pkg/tfshim/sdk-v1/provider.go:97 +0x11c
    github.com/pulumi/pulumi-terraform-bridge/v2/pkg/tfbridge.(*Provider).Update(0xc0009d6380, 0x5a67aa0, 0xc001bc6060, 0xc000396c80, 0xc0009d6380, 0x49bcf01, 0xc0018ded00)
        /home/runner/go/pkg/mod/github.com/pulumi/pulumi-terraform-bridge/v2@v2.11.1-0.20201020163502-64cff1e50894/pkg/tfbridge/provider.go:970 +0x96f
    github.com/pulumi/pulumi/sdk/v2/proto/go._ResourceProvider_Update_Handler.func1(0x5a67aa0, 0xc001bc6060, 0x4dbf7e0, 0xc000396c80, 0x4da7140, 0x9521b20, 0x5a67aa0, 0xc001bc6060)
        /home/runner/go/pkg/mod/github.com/pulumi/pulumi/sdk/v2@v2.12.0/proto/go/provider.pb.go:2280 +0x90
    github.com/grpc-ecosystem/grpc-opentracing/go/otgrpc.OpenTracingServerInterceptor.func1(0x5a67aa0, 0xc001c6b830, 0x4dbf7e0, 0xc000396c80, 0xc000c4b600, 0xc000c4b620, 0x0, 0x0, 0x59e9a80, 0xc0001cd780)
        /home/runner/go/pkg/mod/github.com/grpc-ecosystem/grpc-opentracing@v0.0.0-20180507213350-8e809c8a8645/go/otgrpc/server.go:57 +0x2f2
    github.com/pulumi/pulumi/sdk/v2/proto/go._ResourceProvider_Update_Handler(0x4e76920, 0xc0009d6380, 0x5a67aa0, 0xc001c6b830, 0xc0007bc3c0, 0xc000c10200, 0x5a67aa0, 0xc001c6b830, 0xc000213500, 0x1360)
        /home/runner/go/pkg/mod/github.com/pulumi/pulumi/sdk/v2@v2.12.0/proto/go/provider.pb.go:2282 +0x152
    google.golang.org/grpc.(*Server).processUnaryRPC(0xc0009d61c0, 0x5aaefe0, 0xc001702480, 0xc00005f800, 0xc0009e0b40, 0x94e17f8, 0x0, 0x0, 0x0)
        /home/runner/go/pkg/mod/google.golang.org/grpc@v1.30.0/server.go:1171 +0x511
    google.golang.org/grpc.(*Server).handleStream(0xc0009d61c0, 0x5aaefe0, 0xc001702480, 0xc00005f800, 0x0)
        /home/runner/go/pkg/mod/google.golang.org/grpc@v1.30.0/server.go:1494 +0xcd4
    google.golang.org/grpc.(*Server).serveStreams.func1.2(0xc000de3cf0, 0xc0009d61c0, 0x5aaefe0, 0xc001702480, 0xc00005f800)
        /home/runner/go/pkg/mod/google.golang.org/grpc@v1.30.0/server.go:834 +0xa8
    created by google.golang.org/grpc.(*Server).serveStreams.func1
        /home/runner/go/pkg/mod/google.golang.org/grpc@v1.30.0/server.go:832 +0x20b

    error: update failed

  azure:compute:LinuxVirtualMachine (srm-dev-docker-vm):
    error: transport is closing

  azure:appservice:AppService (srm-dev-app):
    error: transport is closing
```

**Comment by khellang** [comment]
Is this the same issue as these? The stack looks pretty similar... https://github.com/pulumi/pulumi-azure/issues/383 https://github.com/pulumi/pulumi-azure/issues/454

BTW, I'm using Pulumi v2.12.1 and Pulumi.Azure 3.27.0.

**Comment by khellang** [comment]
Huh, just saw there was a new 3.28.0 version released 4 days ago ✨ I tried updating and it looks like it magically solved everything 😅🤷‍♂️

**Comment by andreujuanc** [comment]
It also fails for me:
Callstack: https://pastebin.com/TGWDP7bG
Versions:
```xml
 <PackageReference Include="Pulumi.Azure" Version="3.30.0" />
 <PackageReference Include="Pulumi.AzureAD" Version="2.6.0" />
```
Pulumi cli: v2.13.2

**Comment by nickpainter** [comment]
I am still getting this same error with v3.30.1 of the azure provider.  This reverting to 3.28.0 as suggested by @khellang unfortunately did not work for me.

**Comment by Hawxy** [comment]
For the `web.expandAppServiceLogs` panic, the simplest workaround is to specify the httplogs setting:

ie, for C#:

```csharp
               Logs = new AppServiceLogsArgs
                {
                    HttpLogs = new AppServiceLogsHttpLogsArgs
                    {
                        FileSystem = new AppServiceLogsHttpLogsFileSystemArgs { RetentionInDays = 14, RetentionInMb = 35 }
                    }
                }
```

This problem has been around for a long time now and really needs to be fixed.

**Comment by nickpainter** [comment]
@Hawxy Thank you for providing this workaround.  This is a great way to buy some time before this issue is resolved without needing to resort to a temporary downgrade of the `@pulumi/azure` package.

**Comment by bascoder** [comment]
Is any fix planned for this issue? 
At the moment we are waiting for a fix several weeks now. As a workaround we are thinking of setting app settings using the Azure CLI after the pulumi script as part of our CI/CD pipeline. But we would rather avoid it.

**Comment by mikhailshilkov** [comment]
@bascoder doesn't the workaround above work for you?

**Comment by bascoder** [comment]
> 
> 
> @bascoder doesn't the workaround above work for you?

Do you mean the workaround from Hawxy? 
I'm not just trying to configure app logs. But also a bunch of other app-specific custom settings.

**Comment by mikhailshilkov** [comment]
> Do you mean the workaround from Hawxy?

yes

> I'm not just trying to configure app logs. But also a bunch of other app-specific custom settings.

My understanding is that if you set this property, you can proceed with updating all other settings too.

**Comment by andreujuanc** [comment]
@bascoder Just need to add that Log property to your current AppService description. I did this to a project I'm working on and it works; weeks later, a colleague had this issue and we fixed it with the same workaround.

**Comment by Jmaharman** [comment]
I've had to use the workaround to this problem too. Is this likely to get looked at at some point soon?

**Comment by Jmaharman** [comment]
It should be noted that when a failure takes place, if a resource was created in that same process it will not be logged and but still exists in Azure.

**Comment by kwrightepc** [comment]
> 
> 
> It should be noted that when a failure takes place, if a resource was created in that same process it will not be logged and but still exists in Azure.

Does the Pulumi refresh bring your resource list back in sync at least?

**Comment by Jmaharman** [comment]
I deleted them manually after realising, so no need to run that now.

**Comment by OleAlbers** [comment]
> Does the Pulumi refresh bring your resource list back in sync at least?

Just tested: No. Sadly not (and does not make the `up`-error disappear, either)

**Comment by mikhailshilkov** [comment]
This issue has been fixed in the 4.0.0 version of the provider. Thanks to everyone for their patience and apologies it took so long.
